File: trees/node_distance.cpp
#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* left;
        Node* right;

        Node(int data) {
            this->data = data;
            this->left = NULL;
            this->right = NULL;
        }
};

class Tree {
    public:
        Node* root;

        Tree() {
            this->root = NULL;
        }

        Node* constructTree() {
            Node* root = new Node(8);

            root->left = new Node(3);
            root->left->left = new Node(1);
            root->left->right = new Node(6);
            root->left->right->left = new Node(4);
            root->left->right->right = new Node(7);

            root->right = new Node(12);
            root->right->left = new Node(11);
            root->right->right = new Node(15);
            root->right->right->left = new Node(14);

            return root;
        }

        Node* getLCA(Node* root, int node1, int &node1_height, int node2, int &node2_height, int &dist, int level) {

            Node* temp = NULL;
            if(root == NULL) {
                return NULL;
            }

            if(root->data == node1) {
                node1_height = level;
                temp = root;
            }

            if(root->data == node2) {
                node2_height = level;
                temp = root;
            }

            Node* LCALeft = getLCA(root->left, node1, node1_height, node2, node2_height, dist, level+1);
            Node* LCARight = getLCA(root->right, node1, node1_height, node2, node2_height, dist, level+1);

            if(temp != NULL) {
                return temp;
            }

            // could not LCA in both subtrees
            if(LCALeft == NULL && LCARight == NULL) {
                return NULL;
            // found LCA, return it
            } else if(LCALeft != NULL && LCARight != NULL) {
                dist = node1_height + node2_height - (2*level);
                return root;
            // found LCA in either of the branches, return the one with the LCA
            } else {
                return (LCALeft != NULL) ? LCALeft : LCARight;
            }
        }

        void nodeDistance(Node* root, int node1, int node2) {
            if(root == NULL) {
                return;
            }

            // find LCA
            int node1_height = -1;
            int node2_height = -1;
            int dist = -1;
            Node* lca = getLCA(root, node1, node1_height, node2, node2_height, dist, 0);
            if(lca == NULL || (node1_height == -1) || (node2_height == -1)) {
                cout << "Cannot find LCA" << endl;
            }

            cout << "LCA: " << lca->data << endl;
            cout << "Dist between " << node1 << " and " << node2 << ": " << dist << endl;

            return;
        }

        void printTree(Node* root) {
            if(root == NULL) {
                return;
            }

            printTree(root->left);
            cout << root->data << " ";
            printTree(root->right);
        }

        void destroyTree(Node* root) {
            if(root) {
                destroyTree(root->left);
                destroyTree(root->right);
                delete root;
            }
        }
};

int main() {
    Tree tree;
    tree.root = tree.constructTree();
    Node* root_copy = tree.root;
    tree.nodeDistance(root_copy, 4, 14);
    tree.destroyTree(tree.root);
    return 0;
}

File: trees_heaps/trees_py/node_distance.py
"""
Find the lowest common ancestor of 2 tree nodes
"""

# -*- encoding: utf-8 *-*


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class Tree:
    def __init__(self):
        self.root = None

    def construct_tree(self):
        self.root = Node(8)
        self.root.left = Node(3)
        self.root.left.left = Node(1)
        self.root.left.right = Node(6)
        self.root.left.right.left = Node(4)
        self.root.left.right.right = Node(7)
        self.root.right = Node(12)
        self.root.right.left = Node(11)
        self.root.right.right = Node(15)
        self.root.right.right.left = Node(14)

    def print_level_order_traversal(self, root):
        if root is None:
            return None

        queue = []
        queue.append(root)
        print()
        while(len(queue) > 0):
            count = len(queue)
            while(count > 0):
                curr_node = queue.pop(0)
                print(curr_node.data, end=" ")
                if curr_node.left:
                    queue.append(curr_node.left)
                if curr_node.right:
                    queue.append(curr_node.right)
                count -= 1
            print()

    def find_LCA_util(self, root, val1, val1_height, val2, val2_height, dist, level):
        temp = Node(None)
        if root is None:
            return None

        if root.data == val1:
            val1_height[0] = level
            temp = root

        if root.data == val2:
            val2_height[0] = level
            temp = root

        if val1_height[0] is not -1 and val2_height[0] is not -1 and temp.data:
            return temp

        lca_left = self.find_LCA_util(root.left, val1, val1_height, val2, val2_height, dist, level + 1)
        lca_right = self.find_LCA_util(root.right, val1, val1_height, val2, val2_height, dist, level + 1)

        if temp.data:
            return temp

        if lca_left is None and lca_right is None:
            return None
        elif lca_left is not None and lca_right is not None:
            dist[0] = val1_height[0] + val2_height[0] - (2 * level)
            return root
        elif lca_left is not None:
            return lca_left
        else:
            return lca_right

    def node_distance(self, root, val1, val2):
        if root is None:
            return

        val1_height = [-1]
        val2_height = [-1]
        level = 0
        dist = [-1]
        lca = self.find_LCA_util(root, val1, val1_height, val2, val2_height, dist, level)
        if lca is None or val1_height[0] is -1 or val2_height[0] is -1:
            print("Cannot find LCA")

        print("LCA = ", lca.data)
        print("Dist between ", val1, " and ", val2, " is ", dist[0])


def main():
    tree = Tree()
    tree.construct_tree()
    # tree.print_level_order_traversal(tree.root)
    tree.node_distance(tree.root, 14, 11)
    return


if __name__ == '__main__':
    main()
